### DISCUSSION TITLE
[슬기] assignment 기능구현

### DISCUSSION THREAD

**Discussion by songseul** [Wanted-07-team-9/wanted_assignment_02#12] [Ideas]
### [💙 배포사이트 💙](https://wanted-assignment-02.herokuapp.com)

1.  이슈 목록 화면

    - 이슈 목록 가져오기 API 활용

      ```
        const getPageList = useCallback(async pageNumber => {
      try {
        setIsLoading(true);
        const response = await instance.get(`/issues?sort=comments&page=${pageNumber}`);
        if (response.data.length === 0) {
          setIsNoMore(true);
        }
        setIssueList(prev => [...prev, ...response.data]);
      } catch (err) {
        setIsError(true);
      } finally {
        setIsLoading(false);
      }
      }, []);
      ```

    - Context API를 활용한 API를 연동 하여 깃허브 이슈 목록을 요청
    - useCallback을 활용하여 함수를 메모리제이션 (useEffect 사용시에 불필요한 재랜더링을 줄임)

    ***

    - open 상태의 이슈 중 코멘트가 많은 순으로 정렬

      ```
      const response = await instance.get(`/issues?sort=comments&page=${pageNumber}`);
      ```

    - 요청 url 쿼리 파리미터로 comments가 높은 순으로 정렬

    ***

    - 각 행에는 ‘이슈번호, 이슈제목, 작성자, 작성일, 코멘트수’를 표시

      ```
       <IssueDetailContainer>
        <DetailTitleBox>
          <img src={issueDetail?.user.avatar_url} alt="" />
          <DetailTitle>
            {issueDetail?.title}
            <span>
              작성자: {issueDetail?.name} 작성일: {date}
            </span>
          </DetailTitle>
          <CommentsBox>
            <p>
              comments<span>{issueDetail?.comments} </span>{' '}
            </p>
          </CommentsBox>
        </DetailTitleBox>
        <BodyMarkDown>
          <MarkDown markdown={issueDetail?.body} />{' '}
        </BodyMarkDown>
      </IssueDetailContainer>

      ```

    ***

    - 다섯번째 셀에는 광고 이미지 출력

      ```
                  {index === 4 && (
                    <AdLink href="https://www.wanted.co.kr/">
                      <img src={Wanted} alt="" />
                    </AdLink>
                  )}
      ```

   ![스크린샷 2022-10-30 오후 7 43 45](https://user-images.githubusercontent.com/104307213/198874539-b8c7c7ae-4809-467f-9289-f0707394ac96.png)

 

   - 요청 받은 데이터의 4번째 인덱스에 이미지 광고 이미지 위치
   - 광고 이미지 클릭시 [https://www.wanted.co.kr/](https://www.wanted.co.kr/) 로 이동

   ***

   - 화면을 아래로 스크롤 할 시 이슈 목록 추가 로딩(인피니티 스크롤)

      ```
        const handleObserver = useCallback(
      (entries, observer) => {
        const target = entries[0];
        if (target.isIntersecting && !isLoading) {
          observer.unobserve(bottomLoader.current);
          if (isNoMore) {
            observer.disconnect();
            return;
          }
          nextPageList();
        }
      },
      [isLoading, isNoMore, nextPageList]
      );
      ```

- Intersection Observer API를 사용한 무한 스크롤

  ***

2.  이슈 상세 화면

    ![스크린샷 2022-10-30 오후 7 44 18](https://user-images.githubusercontent.com/104307213/198874549-62323070-2ee9-4c12-9fff-01bec86a5eac.png)

    - 이슈의 상세 내용 표시
    - ‘이슈번호, 이슈제목, 작성자, 작성일, 코멘트 수, 작성자 프로필 이미지, 본문' 표시

3.  공통 헤더

    ![스크린샷 2022-10-30 오후 7 44 45](https://user-images.githubusercontent.com/104307213/198874558-d29690a5-c88d-471b-8d0c-8a0bb5bdb11b.png)

    - 두 페이지는 공통 헤더를 공유합니다.
    - 헤더에는 Organization Name / Repository Name이 표시됩니다.(요청한 데이터의 주소의 유저/레포지토리를 context api 헤더 상태 함수에 저장 )

    ```
      const issueHeader = issueList[0]?.repository_url.split('https://api.github.com/repos/').join('');

       useEffect(() => {
    setHeader(issueHeader);
    }, [issueHeader]);

    ```

    ***
4.  에러화면

  ```
    isError ? (
       <ErrorMessage>
        <SpinnerBox>
          <img src={Spinner} alt="" />
        </SpinnerBox>{' '}
        서버를 확인해 주세요
      </ErrorMessage>
  ```

  - context api 요청시 서버의 문제가 있을때 화면 구현

  ***
### 요구 사항

- 필수 요구 사항

  - [x] 이슈 목록 및 상세 화면 기능 구현
  - [x] Context API를 활용한 API 연동
  - [x] 데이터 요청 중 로딩 표시
  - [x] 에러 화면 구현
  - [x] 지정된 조건(open 상태, 코멘트 많은 순)에 맞게 데이터 요청 및 표시
  - [x] 반응형 웹 구현(UI는 데스크톱, 모바일에서 보았을 때 모두 읽기 편하게 구현)

**Comment by anotheranotherhoon** [comment]
- [x] Assignment3 / 두 페이지 공통 헤더 공유( 헤더의 이름을 함수로 만들어 사용하신 부분이 높은 아이디어를 드리겠습니다. API요청 안에 두 가지 정보가 있다는 생각 까지 닿지 못했었습니다.)

**Comment by im-binary** [comment]
### Assignment1
무한 스크롤 함수의 변수명을 잘 지어 가독성이 좋은 것 같습니다. 또 무한 스크롤 할 때도 로딩 컴포넌트를 적용한 점이 마음에 들어용~!

**Comment by jun-05** [comment]
- 반응형 화면에서 레이아웃이 깨지는 부분이 없는거 같아 좋아요.
- 배포 에러페이지도 잘 처리하신거 같아요. 맞나요? 전 안돼서..

- 에러가 있으면 차라리 page를 이동하는게 좋지 않을까란 생각을 했봤어요. NaN같은게 보이니까요
- 코드를 살짝만 더 적어주시면 보기 편할거 같아요

**Comment by NR0617** [comment]
## 공통헤더를 공유합니다
헤더에서 요청한 데이터의 주소의 유저/레포지토리를 context api 헤더 상태 함수에 저장한 부분이 인상깊었습니다
